### PR TITLE
feat(Table): Ctrl + Shift click keeps previous selection & also shift clicks

### LINKED ISSUE #880: Multiple range selections (with shift and ctrl) are impossible
#### Describe the bug (actual behavior)
When in the table's multi-selection mode, selecting one range of items and then trying to select another one with shift removes the first selection. Shift ignores that ctrl is pressed, nor does shift work when selecting a checkbox column.
image
image

#### Expected behavior
Selecting one range and or single items and then adding another range selection with shift (and ctrl or any other alternative) should keep the selections done before and add just the selected range

##### Steps to reproduce
1. Click on one or more items
1. Use shift on not selected items

#### Additional information
Related to issue #879

### PR DESCRIPTION
**Pull request by r100-stack**
**[CONTINUES]** PR #886 (closed as it was pointing to `main` and not `v2`). Pointing this to `v2` to avoid delaying `v2` due to merge conflicts with `main`.

---

Closes #880 

* Shift click keeps previous selection when holding Ctrl when Shift clicking

**NOTE**: The related issue #879 about `lastSelectedRowId` not updating on checkbox click, will come in a separate PR, hopefully from the external contributor. After that PR, this Ctrl + Shift functionality can be completely realized.

#### Steps

* Click `1.1`
* Shift click `1.2.2`
* Ctrl click `2.1` and `2.3`
* Ctrl + Shift click `1.2.4`

#### New

* Previous selection of `2.1` and `2.3` is preserved

[vokoscreenNG-2022-10-18_16-33-25.webm](https://user-images.githubusercontent.com/45748283/196538489-57f3cef7-52a9-4e5f-860a-61a334ebd274.webm)

#### Old

* Previous selection of `2.1` and `2.3` is lost

[vokoscreenNG-2022-10-18_16-41-20.webm](https://user-images.githubusercontent.com/45748283/196540551-67576324-b5a0-4eb1-b173-18e2e92e78fa.webm)

## Checklist

- [x] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- ~[ ]~ Add component features demo in Storybook (different stories)
- ~[ ]~ Approve test images for new stories
- [x] Add screenshots of the key elements of the component

### CODE REVIEW

File: packages/iTwinUI-react/src/core/Table/Table.test.tsx [hunk 1]
```diff
@@ -352,54 +352,56 @@ it('should handle checkbox clicks', async () => {
   expect(onSelect).toHaveBeenCalledWith([mockedData()[1]], expect.any(Object));
 
   await userEvent.click(checkboxCells[0]);
-  expect(onSelect).toHaveBeenCalledWith(mockedData(), expect.any(Object));
+  expect(onSelect).toHaveBeenCalledWith([], expect.any(Object));
```
Comment (adamhock) [inline packages/iTwinUI-react/src/core/Table/Table.test.tsx:355]: This change must be reverted as it is a result of #864 which is in `main` and not in `v2`.

File: packages/iTwinUI-react/src/core/Table/Table.test.tsx [hunk 2]
```diff
@@ -352,54 +352,56 @@ it('should handle checkbox clicks', async () => {
...
 
   await userEvent.click(checkboxCells[0]);
-  expect(onSelect).toHaveBeenCalledWith([], expect.any(Object));
+  expect(onSelect).toHaveBeenCalledWith(mockedData(), expect.any(Object));
```
Comment (adamhock) [inline packages/iTwinUI-react/src/core/Table/Table.test.tsx:358]: This change must be reverted as well

File: packages/iTwinUI-react/src/core/Table/Table.test.tsx [hunk 3]
```diff
@@ -1853,6 +1895,12 @@ it('should disable row and handle selection accordingly', async () => {
   expect(headerCheckbox.indeterminate).toBe(true);
   expect(headerCheckbox.checked).toBe(false);
 
+  // Deselect all
```
Comment (adamhock) [inline packages/iTwinUI-react/src/core/Table/Table.test.tsx:1898]: This code must be moved back to its original position

**Review by adamhock — COMMENTED** [review]
Looks good to me! Just need to resolve small comments and I will approve

File: packages/iTwinUI-react/src/core/Table/Table.test.tsx (hunk 1 above)
Comment (rohan-kadkol) [reply packages/iTwinUI-react/src/core/Table/Table.test.tsx]: **[DONE]** Ohh, right! Thanks, reverted

File: packages/iTwinUI-react/src/core/Table/Table.test.tsx (hunk 2 above)
Comment (rohan-kadkol) [reply packages/iTwinUI-react/src/core/Table/Table.test.tsx]: **[DONE]** Thanks, reverted

File: packages/iTwinUI-react/src/core/Table/Table.test.tsx (hunk 3 above)
Comment (rohan-kadkol) [reply packages/iTwinUI-react/src/core/Table/Table.test.tsx]: **[DONE]**

**Review by adamhock — APPROVED** [review]
LGTM! 🎉

**Comment by GintV** [comment]
What's the progress on this one? When we could expect this to get released? Will this be included only in 2.x?

cc @ainepi

**Comment by mayank99** [comment]
> What's the progress on this one? When we could expect this to get released? Will this be included only in 2.x?
> 
> cc @ainepi

We put this one on hold to release 2.0, I expect we'll get back to this very soon. I think we would only release it in 2.x.

File: packages/iTwinUI-react/src/core/Table/Table.tsx [hunk 4]
```diff
@@ -622,6 +622,7 @@ export const Table = <
           dispatch({
             type: shiftRowSelectedAction,
             id: row.id,
+            ctrlPressed: event.ctrlKey,
```
Comment (mayank99) [inline packages/iTwinUI-react/src/core/Table/Table.tsx:625]: This doesn't work on macOS with <kbd>cmd</kbd> key. Usually you want to check for [`metaKey`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/metaKey) too.
```suggestion
            ctrlPressed: event.ctrlKey || event.metaKey,
```

And when I was testing this, I found that line 629 also needs to be changed:

```diff
-          (selectionMode === 'single' || !event.ctrlKey)
+          (selectionMode === 'single' || !(event.ctrlKey || event.metaKey))
```

Comment (rohan-kadkol) [reply packages/iTwinUI-react/src/core/Table/Table.tsx:625]: > This doesn't work on macOS with <kbd>cmd</kbd> key. Usually you want to check for [`metaKey`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/metaKey) too.

_(I'm guessing you meant the <kbd>ctrl</kbd> key)_. Thanks, I made that change.

Just curious, what is the Mac equivalent of a Ctrl on Windows/Linux? Like how to do ctrl clicks in Mac's file explorer for example? Using the <kbd>ctrl</kbd> key or the <kbd>command</kbd> key?

Comment (mayank99) [reply packages/iTwinUI-react/src/core/Table/Table.tsx:625]: > > This doesn't work on macOS with cmd key. Usually you want to check for [`metaKey`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/metaKey) too.
> 
> _(I'm guessing you meant the ctrl key)_. Thanks, I made that change.
> 
> Just curious, what is the Mac equivalent of a Ctrl on Windows/Linux? Like how to do ctrl clicks in Mac's file explorer for example? Using the ctrl key or the command key?

no i meant <kbd>cmd</kbd> (or <kbd>command</kbd> or <kbd>⌘</kbd>, all the same). that is the mac equivalent of <kbd>ctrl</kbd> on windows in most cases, including file explorer.

Comment (rohan-kadkol) [reply packages/iTwinUI-react/src/core/Table/Table.tsx]: **[DONE]** Ohh, makes sense, thanks. I made the change in the two places `event.ctrlPressed` was used ([Commits](https://github.com/iTwin/iTwinUI-react/pull/888/files/f37c20902961eb76118aa148f7a0d21f467688c2..3ca38c2439a04451eb9682099b198db22d97dcfa))

**Review by mayank99 — APPROVED** [review]

**Review by mayank99 — APPROVED** [review]

**Pull request merged**